plugins/workspace-plugin/src/workspace-projects-manager.ts

```diff
@@ -33,37 +34,56 @@ export function handleWorkspaceProjects(pluginContext: theia.PluginContext, proj
   });
 }
 
+type Project = cheApi.workspace.devfile.Project | cheApi.workspace.ProjectConfig;
+
 abstract class WorkspaceProjectsManager {
-  watchers: theia.FileSystemWatcher[] = [];
+  protected watchers: theia.FileSystemWatcher[] = [];
+  protected workspaceFolderUpdater = new WorkspaceFolderUpdater();
 
   constructor(protected pluginContext: theia.PluginContext, protected projectsRoot: string) {}
 
+  abstract getProjectPath(project: Project): string;
+  abstract getProjects(workspace: cheApi.workspace.Workspace): Project[];
   abstract async selectProjectToCloneCommands(workspace: cheApi.workspace.Workspace): Promise<TheiaImportCommand[]>;
   abstract async updateOrCreateProject(workspace: cheApi.workspace.Workspace, projectFolderURI: string): Promise<void>;
   abstract deleteProject(workspace: cheApi.workspace.Workspace, projectFolderURI: string): void;
 
   async run(workspace?: cheApi.workspace.Workspace): Promise<void> {
-    if (!theia.workspace.name) {
-      // no workspace opened, so nothing to clone / watch
-      return;
-    }
-
     if (!workspace) {
       workspace = await che.workspace.getCurrentWorkspace();
     }
-    const cloneCommandList = await this.selectProjectToCloneCommands(workspace);
-    await this.executeCloneCommands(cloneCommandList);
 
+    const cloningPromise = this.executeCloneCommands(workspace);
+
+    theia.window.withProgress({ location: { viewId: 'explorer' } }, () => cloningPromise);
+
+    await cloningPromise;
     await this.startSyncWorkspaceProjects();
```

Comment:
I don't thingk "startSyncWorkspaceProjects" is really async. Let's get rid of async/await.

Reply:
I create the issue for the refactoring workspace creation related logic: https://github.com/eclipse/che/issues/19082